src/main/drivers/serial.c

```diff
@@ -37,13 +37,16 @@ uint32_t serialGetBaudRate(serialPort_t *instance)
 
 void serialWrite(serialPort_t *instance, uint8_t ch)
 {
-    instance->vTable->serialWrite(instance, ch);
+    while (!serialTxBytesFree(instance)) {
+    };
+
+	instance->vTable->serialWrite(instance, ch);
 }
```

Comment:
I did implement circular buffering for VCP some time ago. It uses same buffering as for serial and implements 'chunk' interface between serial and VCP part. Maybe it is worth reviving ... 

https://github.com/cleanflight/cleanflight/pull/886/files